sdk/core/Azure.Core/src/DynamicData/DynamicData.AllowList.cs

```diff
@@ -0,0 +1,202 @@
+            }
+
+            private static bool IsAllowedCollectionType(Type type)
+            {
+                return
+                    IsAllowedListType(type) ||
+                    IsAllowedDictionaryType(type);
+            }
+
+            private static bool IsAllowedListType(Type type)
+            {
+                if (!type.IsGenericType)
+                {
+                    return false;
+                }
+
+                if (type.GetGenericTypeDefinition() != typeof(List<>))
+                {
+                    return false;
+                }
+
+                Type[] types = type.GetGenericArguments();
+                return IsAllowedType(types[0]);
+            }
+
+            private static bool IsAllowedDictionaryType(Type type)
+            {
+                if (!type.IsGenericType)
+                {
+                    return false;
+                }
+
+                if (type.GetGenericTypeDefinition() == typeof(Dictionary<,>))
+                {
+                    Type[] types = type.GetGenericArguments();
+                    if (IsAllowedType(types[0]) && IsAllowedType(types[1]))
+                    {
+                        return true;
+                    }
+                }
+
+                return false;
+            }
+
+            private static bool IsAllowedEnumerableType(Type type)
+            {
+                if (!type.IsGenericType)
+                {
+                    return false;
+                }
+
+                if (!type.IsInterface)
+                {
+                    return false;
+                }
+
+                if (type.GetGenericTypeDefinition() != typeof(IEnumerable<>))
+                {
+                    return false;
+                }
+
+                Type[] types = type.GetGenericArguments();
+                return IsAllowedType(types[0]);
+            }
+
+            private static bool IsAllowedPocoType(Type type, HashSet<Type> ancestorTypes)
+            {
+                if (!HasPublicParameterlessConstructor(type) && !IsAnonymousType(type))
+                {
+                    return false;
+                }
+
+                foreach (PropertyInfo property in type.GetProperties())
+                {
+                    if (!property.CanRead)
+                    {
+                        return false;
+                    }
+
+                    if (!property.CanWrite && !IsAnonymousType(type))
+                    {
+                        return false;
+                    }
+
+                    if (IsAllowedKnownType(property.PropertyType))
+                    {
+                        continue;
+                    }
+
+                    // Trust but verify
+                    if (ancestorTypes.Contains(property.PropertyType))
+                    {
+                        continue;
+                    }
+
+                    // Recurse
+                    ancestorTypes.Add(type);
```

Comment:
It can't return `true` if it continues, so the method will eventually return false if a type added to the list doesn't pass `IsAllowedPocoType`.  I'm happy to add any tests cases you'd like to verify this.

Reply:
Per offline discussion, will fix.

Reply:
Actually, when I looked at modifying this, I remembered why I did it this way, and I'm not convinced yet that it doesn't work.  The goal here is to break cycles in the object graph.  It does continue past a given property if it's already seen its as an ancestor.  But because it iterates over all properties of the type, it will check all of them and either allow the type because it doesn't have any non-allowed properties (just a weird POCO with a cycle in it), or not allow it because it has properties with non-allowed types.

I have [tests for this here](https://github.com/Azure/azure-sdk-for-net/pull/36791/files#diff-4cc147e6d7f7fc4db981f75da3aa763b71ca9c7cbb2871f15dc558f8c1cc78ceR126-R248). 

Per the comment about HashSet, I will still look into a more efficient way to track ancestors, but since you can't walk up a Type branch from a given property, I think we do have to keep some sort of collection of ancestors to recurse on.

Reply:
Ah, OK. I think your code works. It might first let a type through the filter (say that it's allowed), but then later catch it when it finally gets to a property that is not allowed.